### ISSUE TITLE
Device mismatch in matrix multiplication leads to "RuntimeError: CUDA error: an illegal memory access was encountered"

### ISSUE THREAD

**Issue by erikbrntsn** [pytorch/pytorch#43700]
## 🐛 Bug

I encountered the error "RuntimeError: CUDA error: an illegal memory access was encountered" (maybe related to https://github.com/pytorch/pytorch/issues/21819 or parts of it) which caused me quite some confusion.

The problem is seen on pytorch 1.6.0
I have also tried on '1.5.0+cu92' where a more informative error message is provided: "RuntimeError: Expected object of device type cuda but got device type cpu for argument #2 'mat2' in call to _th_mm"

Below are given code snippets to reproduce the problem.
The must frustrating part about this problem is that in my original setup the "..  illegal memory access ..." error happened on a entirely different line than those lines that seem to be the root cause. I can try to produce a minimal example where the error occurs in a seemingly unrelated part of the program if needed.

The code below is a minimal example I managed to construct where the problem consistently occurs on my system.

## To Reproduce

Steps to reproduce the  #behavior:
Run the following code
```
import torch

s = torch.rand(size=(3, 4)).to("cuda")

s_new = torch.zeros(s.shape)

s_new[:, :-1] = s[:, :-1].matmul(s_new[:, :-1])
```
or related I assume
```
import torch

s = torch.rand(size=(3, 4)).to("cuda")

s_new = torch.zeros(s.shape)

print(s[:, :-1].matmul(s_new[:, :-1]))
```
which instead results in "RuntimeError: copy_if failed to synchronize: cudaErrorIllegalAddress: an illegal memory access was encountered"

The problem is not seen if `s_new = torch.zeros(s.shape)` is replaced with `s_new = torch.zeros(s.shape, device=s.device)`

Note: the same error happens if .matmul is substituted with .mm or @.
Note: if instead of matrix multiplication + or * is used a device mismatch error is provided.

## Expected behavior

An error message related to device mismatch is provided

## Environment

PyTorch version: 1.6.0
Is debug build: False
CUDA used to build PyTorch: 10.2
OS: Microsoft Windows 10 Enterprise
GCC version: Could not collect
Clang version: Could not collect
CMake version: version 3.14.5
Python version: 3.7 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 10.1.243
GPU models and configuration: GPU 0: GeForce RTX 2080 Ti
Nvidia driver version: 441.22
cuDNN version: Could not collect
Versions of relevant libraries:
[pip3] numpy==1.18.2
[pip3] torch==1.6.0
[pip3] torchvision==0.7.0
[conda] Could not collect

cc @ngimel

**Comment by glaringlee** [comment]
@erikbrntsn 
Thanks for filing this issue. 

mm (matmul call it) is ported from Legacy TH to ATen in 1.6 and the new mm api doesn't check whether all input tensors have same device, that's why you get 'illegal memory access',  we will fix this.

**Comment by ngimel** [comment]
This is fixed in #43505, closing, please reopen if problem persists.
